src/sqlfluff/rules/L050.py

```diff
@@ -0,0 +1,115 @@
+       :force:
+
+        ^
+
+        SELECT
+            a
+        FROM foo
+
+        -- Beginning on an indented line is also forbidden,
+        -- (the • represents space).
+
+        ••••SELECT
+        ••••a
+        FROM
+        ••••foo
+
+    | **Best practice**
+    | Start file on either code or comment, the ^ represents the beginning of file.
+
+    .. code-block:: sql
+       :force:
+
+
+        ^SELECT
+            a
+        FROM foo
+
+        -- Including an initial block comment.
+
+        ^/*
+        This is a description of my SQL code.
+        */
+        SELECT
+            a
+        FROM
+            foo
+
+        -- Including an initial inline comment.
+
+        ^--This is a description of my SQL code.
+        SELECT
+            a
+        FROM
+            foo
+    """
+
+    @staticmethod
+    def _potential_template_collision(context: RuleContext) -> bool:
+        """Check for any templated raw slices that intersect with source slices in the raw_stack.
+
+        Returns:
+            :obj:`bool` indicating a preceding templated raw slice has been detected.
+
+        """
+        templated_file = context.segment.pos_marker.templated_file
+        for segment in context.raw_stack:
+            source_slice = segment.pos_marker.source_slice
+            raw_slices = templated_file.raw_slices_spanning_source_slice(source_slice)
+            if any(
+                raw_slice
+                for raw_slice in raw_slices
+                if raw_slice.slice_type == "templated"
+            ):
+                return True
+
+        return False
+
+    def _eval(self, context: RuleContext) -> Optional[LintResult]:
+        """Files must not begin with newlines or whitespace."""
+        # If parent_stack is empty we are currently at FileSegment.
+        if len(context.parent_stack) == 0:
+            return None
+
+        # If the current segment is either comment or code and all
+        # previous segments are forms of whitespace then we can
+        # remove these earlier segments.
+        # Given the tree stucture, we make sure we are at the
+        # first leaf to avoid repeated detection.
+        if (
+            (context.segment.is_comment or context.segment.is_code)
+            & (len(context.raw_stack) > 0)
+            & set(segment.name for segment in context.raw_stack).issubset(
+                {"newline", "whitespace", "Dedent"}
+            )
+            & (not context.segment.is_expandable)
+        ):
+            # It is possible that a template segment (e.g. {{ config(materialized='view') }})
+            # renders to an empty string and as such is omitted from the parsed tree.
+            # We therefore should flag if a templated raw slice intersects with the
+            # source slices in the raw stack and skip this rule to avoid risking
+            # collisions with template objects.
+            if self._potential_template_collision(context):
+                return None
```

Comment:
Should this be outside the `if` as an early exit? Or is this more computational expensive and/or less likely to be hit so better to check we're in the `if` statement first before hitting this?

Reply:
It relies on the current context being the first code/comment segment, which is caught by the outer if block. It can then do a simple check back over the raw stack to see if there are any overlapping templated raw slices. Putting it outside the block would require more complex checks/more hits, so definitely worth keeping it here.